pkg/controller/validation/check_snapshot.go

```diff
@@ -16,32 +17,46 @@ import (
 var log = logf.Log.WithName("check_snapshot")
 
 const (
-	snapshotCRDVersion = "v1alpha1"
-	snapshotCRDName    = "volumesnapshots.snapshot.storage.k8s.io"
+	alphaVersion           = "v1alpha1"
+	snapshotCRDName        = "volumesnapshots.snapshot.storage.k8s.io"
+	snapshotClassCRDName   = "volumesnapshotclasses.snapshot.storage.k8s.io"
+	snapshotContentCRDName = "volumesnapshotcontents.snapshot.storage.k8s.io"
 )
 
-// Returns an error if the v1alpha1 VolumeSnapshot CRD is installed in the cluster
+// Returns an error if the v1alpha1 CRD is installed in the cluster
 // and nil otherwise. This is used to prevent upgrading where clusters have manually
 // installed versions we don't support
+
 func CheckAlphaSnapshot(client client.Client) error {
-	snapshotCRD := &apiextv1beta1.CustomResourceDefinition{}
-	err := client.Get(context.TODO(), types.NamespacedName{Name: snapshotCRDName, Namespace: corev1.NamespaceAll}, snapshotCRD)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			// If we can't find the VolumeSnapshot CRD, then we're safe to proceed
-			return nil
+	crdMap := map[string]string{"VolumeSnapshot": snapshotCRDName,
+		"VolumeSnapshotClass":   snapshotClassCRDName,
+		"VolumeSnapshotContent": snapshotContentCRDName}
+	foundCRD := []string{}
+	crd := &apiextv1beta1.CustomResourceDefinition{}
+	for k, v := range crdMap {
+		err := client.Get(context.TODO(), types.NamespacedName{Name: v, Namespace: corev1.NamespaceAll}, crd)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				errMsg := "Error attempting to obtain existing " + k + " CRD."
+				log.Error(err, errMsg)
+				return err
+			}
+			log.V(5).Info("CRD for " + k + " not found.")
+			continue
 		}
-		log.Error(err, "Error attempting to obtain existing VolumeSnapshot CRD")
-		return err
-	}
 
-	for _, version := range snapshotCRD.Spec.Versions {
-		if version.Name == snapshotCRDVersion {
-			err = apierrors.NewConflict(schema.GroupResource{Resource: "VolumeSnapshot"},
-				"v1alpha1 VolumeSnapshot installed.", nil)
-			log.Error(err, "Unable to update cluster as VolumeSnapshot v1alpha1 is detected. Remove this version to allow the upgrade to proceed.")
-			return err
+		for _, version := range crd.Spec.Versions {
+			if version.Name == alphaVersion {
+				foundCRD = append(foundCRD, v)
+				log.Error(err, "v1alpha1 "+k+" installed")
+			}
 		}
 	}
+	if len(foundCRD) > 0 {
+		errString := "Unable to update cluster as v1alpha1 version of " + strings.Join(foundCRD, ", ") + "is detected. Remove these CRDs to allow the upgrade to proceed."
+		err := apierrors.NewConflict(schema.GroupResource{Resource: "Cluster Storage Operator"},
```

Comment:
Why API ConflictError? That says an Update/Patch did not succeed because someone has updated the object already. I'd prefer a new error type.

Reply:
I assumed Conflict because it said:

> the requested operation cannot be completed due to a conflict in the operation.

I thought the Conflict would be attempting to install the CRDs while a conflicting CRD (v1alpha1) was already installed. If I want to create a new error type, then I'd need to update the k8s.io/apimachinery repo? And call it `VersionConflict` ?

Reply:
No, just use a new go error type in check_snapshot.go, you don't need to be limited to API errors.

Reply:
I created a new error, `AlphaVersionError`, and updated the clusterstorage_controller to check for this type of error for validation purposes.